pennylane/optimize/shot_adaptive.py

```diff
@@ -267,6 +267,78 @@ def cost(*args, **kwargs):
 
         return [np.concatenate(i) for i in zip(*grads)]
 
+    @staticmethod
+    def qnode_weighted_random_sampling(qnode, shots, argnums, *args, **kwargs):
+        """Returns an array of length ``shots`` containing single-shot estimates
+        of the Hamiltonian gradient. The shots are distributed randomly over
+        the terms in the Hamiltonian, as per a multinomial distribution.
+
+        Args:
+            qnode (.QNode): A QNode that returns the expectation value of a Hamiltonian.
+            shots (int): The number of shots used to estimate the Hamiltonian expectation
+                value. These shots are distributed over the terms in the Hamiltonian,
+                as per a Multinomial distribution.
+            argnums (Sequence[int]): the QNode argument indices which are trainable
+            *args: Arguments to the QNode
+            **kwargs: Keyword arguments to the QNode
+
+        Returns:
+            array[float]: the single-shot gradients of the Hamiltonian expectation value
+        """
+        qnode.construct(args, kwargs)
+        base_func = qnode.func
+        tape = qnode.tape
+        [expval] = tape.measurements
+        coeffs, observables = expval.obs.terms()

```

Comment:
```suggestion
        coeffs, observables = expval.obs.terms() if instance(expval.obs, qml.Hamiltonian) else (1.0, expval.obs)

```
If we're exposing this function as a public method then it's probably worth casting the observable as a `Hamiltonian` first.